gymnasium/dev_wrappers/lambda_action.py

```diff
@@ -0,0 +1,181 @@
+
+    Vectorized environment:
+        >>> env = gymnasium.vector.make("CarRacing-v2", continuous=False, num_envs=2)
+        >>> env = LambdaActionV0(
+        ...     env, lambda action, _: action.astype(np.int32), [None for _ in range(2)]
+        ... )
+        >>> obs, rew, term, trunc, info = env.step([np.float64(1.2), np.float64(1.2)])
+    """
+
+    def __init__(
+        self,
+        env: gymnasium.Env,
+        func: Callable,
+        args: FuncArgType[Any],
+        action_space: Optional[gymnasium.Space] = None,
+    ):
+        """Initialize LambdaAction.
+
+        Args:
+            env (Env): The gymnasium environment
+            func (Callable): function to apply to action
+            args: function arcuments
+            action_space: wrapped environment action space
+        """
+        super().__init__(env)
+
+        self.func = func
+        self.func_args = args
+        if action_space is None:
+            self.action_space = env.action_space
+        else:
+            self.action_space = action_space
+
+    def action(self, action):
+        """Apply function to action."""
+        return apply_function(self.action_space, action, self.func, self.func_args)
+
+    def _transform_space(self, env: gymnasium.Env, args: FuncArgType):
+        """Process the space and apply the transformation."""
+        return transform_space_bounds(env.action_space, args, transform_space_bounds)
+
+
+class ClipActionsV0(LambdaActionV0):
+    """A wrapper that clips actions passed to :meth:`step` with an upper and lower bound.
+
+    Basic Example:
+        >>> import gymnasium
+        >>> from gymnasium.wrappers import ClipActionsV0
+        >>> env = gymnasium.make("BipedalWalker-v3")
+        >>> env.action_space
+        Box(-1.0, 1.0, (4,), float32)
+        >>> env = ClipActionsV0(env, (-0.5, 0.5))
+        >>> env.action_space
+        Box(-0.5, 0.5, (4,), float32)
+
+    Clip with only a lower or upper bound:
+        >>> env = gymnasium.make('CarRacing-v1')
+        >>> env.action_space
+        Box([-1.  0.  0.], 1.0, (3,), float32)
+        >>> env = ClipActionsV0(env, (None, 0.5))
+        >>> env.action_space
+        Box([-1.  0.  0.], 0.5, (3,), float32)
+
+    Composite action space example:
+        >>> env = ExampleEnv(action_space=Dict(body=Dict(head=Box(0.0, 10.0, (1,))), left_arm=Discrete(4), right_arm=Box(0.0, 5.0, (1,))))
+        >>> env.action_space
+        Dict(body: Dict(head: Box(0.0, 10.0, (1,), float32)), left_arm: Discrete(4), right_arm: Box(0.0, 5.0, (1,), float32))
+        >>> args = {"right_arm": (0, 2), "body": {"head": (0, 3)}}
+        >>> env = ClipActionsV0(env, args)
+        >>> env.action_space
+        Dict(body: Dict(head: Box(0.0, 3.0, (1,), float32)), left_arm: Discrete(4), right_arm: Box(0.0, 2.0, (1,), float32))
+    """
+
+    def __init__(
+        self, env: gymnasium.Env, args: FuncArgType[typing.Tuple[float, float]]
```

Comment:
Could we clip with jp.ndarrays (e.g., different min/max for each element of Box space)?

Reply:
Actually yes, if you mean this:

```python
>>> from gym.wrappers import ClipActionsV0
>>> from tests.dev_wrappers.utils import TestingEnv
>>> from gym.spaces import Box
i>>> mport numpy as np

>>> env = TestingEnv(action_space=Box(0, 5, (3,)))
>>> env.action_space
Box(0.0, 5.0, (3,), float32)

>>> wrapped_env = ClipActionsV0(env, (np.array([0,0,0]), np.array([2,3,1])))
>>> wrapped_env.action_space
Box(0.0, [2. 3. 1.], (3,), float32)

>>> wrapped_env.step(np.array([10,10,10]))
(array(-6.842204, dtype=float32), 0, False, False, {'action': array([2, 3, 1])})
```

Reply:
There is still an open discussion about type hints, but considering also this way of clipping typing should be changed